core/src/services/huggingface/core.rs

```diff
@@ -0,0 +1,182 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::fmt::Debug;
+
+use bytes::Bytes;
+use http::Request;
+use http::Response;
+use http::{header, StatusCode};
+use serde_json::json;
+
+use super::backend::RepoType;
+use super::error::parse_error;
+use crate::raw::*;
+use crate::*;
+
+pub struct HuggingFaceCore {
+    pub repo_type: RepoType,
+    pub repo_id: String,
+    pub revision: String,
+    pub root: String,
+    pub read_token: Option<String>,
+
+    pub client: HttpClient,
+}
+
+impl Debug for HuggingFaceCore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("HuggingFaceCore")
+            .field("repo_type", &self.repo_type)
+            .field("repo_id", &self.repo_id)
+            .field("revision", &self.revision)
+            .field("root", &self.root)
+            .finish_non_exhaustive()
+    }
+}
+
+impl HuggingFaceCore {
+    pub async fn hf_path_info(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let p = build_abs_path(&self.root, path)
+            .trim_end_matches('/')
+            .to_string();
+
+        let url = match self.repo_type {
+            RepoType::Model => format!(
+                "https://huggingface.co/api/models/{}/paths-info/{}",
+                &self.repo_id, &self.revision
+            ),
+            RepoType::Dataset => format!(
+                "https://huggingface.co/api/datasets/{}/paths-info/{}",
+                &self.repo_id, &self.revision
+            ),
+        };
+
+        let mut req = Request::post(&url);
+
+        if let Some(token) = &self.read_token {
+            let auth_header_content = format!("Bearer {}", token);
+            req = req.header(header::AUTHORIZATION, auth_header_content);
+        }
+
+        req = req.header(header::CONTENT_TYPE, "application/x-www-form-urlencoded");
+
+        let path = percent_encode_path(&p);
+        let req_body = &json!({
+            "paths": path,
+            "expand": "True",
+        });
+
+        // NOTE: We need to transfer body to x-www-form-urlencoded format.
+        let req_body =
+            serde_urlencoded::to_string(&req_body).expect("failed to encode request body");
```

Comment:
Can we avoid introducing `serde_urlencoded`?

Reply:
Dropped `serde_urlencoded` and `json!`. Right now, I found that `format!("paths={}&expand=True", percent_encode_path(&p));` will do the job just fine.